README
NAME
    any::feature - Backwards-compatible handling of new syntactic features

VERSION
    version 1.100840

SYNOPSIS
        use any::feature 'say';
        say 'Hello, world!';

DESCRIPTION
  THE PROBLEM
    Perl 5.10 introduces new syntactic features which you can activate and
    deactivate with the "feature" module. You want to use the "say" feature
    in a program that's supposed to run under both Perl 5.8 and 5.10. So
    your program looks like this:

        use feature 'say';
        say 'Hello, world!';

    But this only works in Perl 5.10, because there is no "feature" module
    in Perl 5.8. So you write

        use Perl6::Say;
        say 'Hello, world!';

    This works, but it's strange to force Perl 5.10 users to install
    Perl6::Say when the "say" feature is included in Perl 5.10.

  THE SOLUTION
    Use "any::feature"!

    WARNING: This is just a proof-of-concept.

    "any::feature" can be used like Perl 5.10's "feature" and will try to
    "do the right thing", regardless of whether you use Perl 5.8 or Perl
    5.10.

    At the moment, this is just a proof-of-concept and only handles the
    "say" feature. If things work out, I plan to extend it with other Perl
    5.10 features.

    The following programs should work and exhibit the same behaviour both
    in Perl 5.8 and Perl 5.10.

    This program will work:

        use any::feature 'say';
        say 'Hello, world!';

    This program will fail at compile-time:

        use any::feature 'say';
        say 'Hello, world!';

        no any::feature 'say';
        say 'Oops';

    The features are lexically scoped, which is how they work in Perl 5.10:

        {
            use any::feature 'say';
            say 'foo';
        }
        say 'bar';     # dies at compile-time

FUNCTIONS
  dispatch
    Takes as arguments a direction ("activate" or "deactivate"), a package
    name and a feature name. Activates or deactivates the given feature for
    the given package.

  activate
    Takes as arguments a package name and a feature name. Uses "dispatch()"
    to activate the given feature in the given package.

  deactivate
    Takes as arguments a package name and a feature name. Uses "dispatch()"
    to deactivate the given feature in the given package.

  import
    Takes the same arguments as Perl 5.10's "use feature" pragma. Uses
    "activate()" and "deactivate()" to do its job.

  unimport
    Takes the same arguments as Perl 5.10's "no feature" pragma. Uses
    "activate()" and "deactivate()" to do its job.

  get_effective_version
    Uses Perl::Version to get the version number of the current perl
    interpreter. This is used to decide the course of action.

  get_effective_revision
    Uses Perl::Version to get the revision number of the current perl
    interpreter. This is used to decide the course of action.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=any-feature>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/any-feature/>.

    The development version lives at
    <http://github.com/hanekomu/any-feature/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2009 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension any-feature

1.100840  2010-03-25 20:59:46 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.05  Tue 2010.02.09 23:59:49 CET (Marcel Gruenauer <marcel@cpan.org>)
     - releng

0.04  Thu 2009.10.08 22:29:11 CEST (Marcel Gruenauer <marcel@cpan.org>)
     - re-release with fixed inc::Module::Install::ReadmeMarkdownFromPod

0.03 Tue 2009.10.06 12:34:18 CEST (Marcel Gruenauer <marcel@cpan.org>)
     - use mysubs.pm instead of B::Hooks::EndOfScope and Sub::Delete for safer
       lexical subs (thanks chocolateboy)
     - use Capture::Tiny instead of IO::Capture::* in tests

0.02 Mon Oct  5 09:53:49 CEST 2009 (Marcel Gruenauer <marcel@cpan.org>)
     - UNIVERSAL::require is now used for 5.10.* as well
     - B::Hooks::Scope is only loaded for 5.8.*
     - added INSTALL
     - better README, README.mkdn

0.01 Wed Feb 18 13:42:36 CET 2009 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/any/feature.pm
use 5.008;
use strict;
use warnings;

package any::feature;
# ABSTRACT: Backwards-compatible handling of new syntactic features

use Perl::Version;
use Carp qw(croak);
use UNIVERSAL::require;

BEGIN {
    if ($] =~ /^5\.008/) {
        require mysubs;
    }
}
our $VERSION = '0.05';
my %dispatch = (
    activate => {
        say => {
            8 => sub {
                Perl6::Say->require or die $@;

                # no need to specify the target - mysubs uses
                # Devel::Pragma::ccstash, which does the right thing if this
                # module is subclassed
                mysubs->import(say => \&Perl6::Say::say);
            },
            10 => sub {
                feature->require or die $@;
                feature->import(qw(say));
            },
        },
    },
    deactivate => {
        say => {
            8 => sub {

                # no need to specify the target - mysubs uses
                # Devel::Pragma::ccstash, which does the right thing if this
                # module is subclassed
                mysubs->unimport('say');
            },
            10 => sub {
                feature->require or die $@;
                feature->unimport(qw(say));
            },
        },
    },
);

sub get_effective_revision {
    our $perl_version ||= Perl::Version->new($]);
    $perl_version->revision;
}

sub get_effective_version {
    our $perl_version ||= Perl::Version->new($]);
    my $version = $perl_version->version;

    # assume perl 5.11 and later behave the same as perl 5.10
    $version = 10 if $version > 10;
    $version;
}

sub dispatch {
    my ($direction, $target, $feature) = @_;
    my $code = $dispatch{$direction}{$feature}{ get_effective_version() };
    unless (ref $code eq 'CODE') {
        croak "unknown feature '$feature'\n";
    }
    $code->($target);
}

sub activate {
    my ($target, $feature) = @_;
    dispatch(activate => $target, $feature);
}

sub deactivate {
    my ($target, $feature) = @_;
    dispatch(deactivate => $target, $feature);
}

sub import {
    my $pkg    = shift;
    my $target = (caller());
    $Carp::CarpInternal{__PACKAGE__}++;
    unless (get_effective_revision() == 5) {
        croak "perl 5.* required\n";
    }
    for my $feature (@_) {
        activate($target, $feature);
    }
}

sub unimport {
    my $pkg    = shift;
    my $target = (caller());
    unless (get_effective_revision() == 5) {
        croak "perl 5.* required\n";
    }
    for my $feature (@_) {
        deactivate($target, $feature);
    }
}
1;


__END__
=pod

=head1 NAME

any::feature - Backwards-compatible handling of new syntactic features

=head1 VERSION

version 1.100840

=head1 SYNOPSIS

    use any::feature 'say';
    say 'Hello, world!';

=head1 DESCRIPTION

=head2 THE PROBLEM

Perl 5.10 introduces new syntactic features which you can activate and
deactivate with the C<feature> module. You want to use the C<say> feature in a
program that's supposed to run under both Perl 5.8 and 5.10. So your program
looks like this:

    use feature 'say';
    say 'Hello, world!';

But this only works in Perl 5.10, because there is no C<feature> module in
Perl 5.8. So you write

    use Perl6::Say;
    say 'Hello, world!';

This works, but it's strange to force Perl 5.10 users to install L<Perl6::Say>
when the C<say> feature is included in Perl 5.10.

=head2 THE SOLUTION

Use C<any::feature>!

WARNING: This is just a proof-of-concept.

C<any::feature> can be used like Perl 5.10's C<feature> and will try to "do
the right thing", regardless of whether you use Perl 5.8 or Perl 5.10.

At the moment, this is just a proof-of-concept and only handles the C<say>
feature. If things work out, I plan to extend it with other Perl 5.10
features.

The following programs should work and exhibit the same behaviour both in Perl
5.8 and Perl 5.10.

This program will work:

    use any::feature 'say';
    say 'Hello, world!';

This program will fail at compile-time:

    use any::feature 'say';
    say 'Hello, world!';

    no any::feature 'say';
    say 'Oops';

The features are lexically scoped, which is how they work in Perl 5.10:

    {
        use any::feature 'say';
        say 'foo';
    }
    say 'bar';     # dies at compile-time

=head1 FUNCTIONS

=head2 dispatch

Takes as arguments a direction (C<activate> or C<deactivate>), a package name
and a feature name. Activates or deactivates the given feature for the given
package.

=head2 activate

Takes as arguments a package name and a feature name. Uses C<dispatch()> to
activate the given feature in the given package.

=head2 deactivate

Takes as arguments a package name and a feature name. Uses C<dispatch()> to
deactivate the given feature in the given package.

=head2 import

Takes the same arguments as Perl 5.10's C<use feature> pragma. Uses
C<activate()> and C<deactivate()> to do its job.

=head2 unimport

Takes the same arguments as Perl 5.10's C<no feature> pragma. Uses
C<activate()> and C<deactivate()> to do its job.

=head2 get_effective_version

Uses L<Perl::Version> to get the version number of the current perl
interpreter.  This is used to decide the course of action.

=head2 get_effective_revision

Uses L<Perl::Version> to get the revision number of the current perl
interpreter.  This is used to decide the course of action.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=any-feature>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/any-feature/>.

The development version lives at
L<http://github.com/hanekomu/any-feature/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



inc/IO/Capture/Stderr.pm
#line 1
package IO::Capture::Stderr;
use strict;
use warnings;
use Carp;
use base qw/IO::Capture/;
use IO::Capture::Tie_STDx;

sub _start {
	my $self = shift;
	$self->line_pointer(1);

	if ( _capture_warn_check() ) {
		$self->{'IO::Capture::handler_save'} = defined $SIG{__WARN__} ? $SIG{__WARN__} : 'DEFAULT';
		$SIG{__WARN__} = sub {print STDERR @_;};
	}
	else {
		$self->{'IO::Capture::handler_save'} = undef;
	}
    tie *STDERR, "IO::Capture::Tie_STDx";
}

sub _retrieve_captured_text {
    my $self = shift;
    my $messages = \@{$self->{'IO::Capture::messages'}};

     @$messages = <STDERR>;
	return 1;
}

sub _check_pre_conditions {
	my $self = shift;

	return unless $self->SUPER::_check_pre_conditions;

	if (tied *STDERR) {
		carp "WARNING: STDERR already tied, unable to capture";
		return;
	}
	return 1;
}

sub _stop {
	my $self = shift;
    untie *STDERR;
	$SIG{__WARN__} = $self->{'IO::Capture::handler_save'} if defined $self->{'IO::Capture::handler_save'};
    return 1;
}

#  _capture_warn_check
#
#  Check to see if SIG{__WARN__} handler should be set to direct output
# from warn() to IO::Capture::Stderr.  
#   There are three things to take into consideration.  
#   
#   1) Is the version of perl less than 5.8?
#      - Before 5.8, there was a bug that caused output from warn() 
#        not to be sent to STDERR if it (STDERR) was tied.
#        So, we need to put a handler in to send warn() text to
#        STDERR so IO::Capture::Stderr will capture it.
#   2) Is there a handler set already?
#      - The default handler for SIG{__WARN__} is to send to STDERR.
#        But, if it is set by the program, it may do otherwise, and
#        we don't want to break that. 
#   3)  FORCE_CAPTURE_WARN => 1
#      - To allow users to override a previous handler that was set on
#        SIG{__WARN__}, there is a variable that can be set.  If set,
#        when there is a handler set on IO::Capture::Stderr startup,
#        it will be saved and a new hander set that captures output to
#        IO::Capture::Stderr.  On stop, it will restore the programs
#        handler.
#      
#
#                    
#    Perl   |  FORCE_CAPTURE_WARN  |  Program has   | Set our own
#    < 5.8  |  is set              |  handler set   | handler
#   --------+----------------------+----------------+------------
#           |                      |                |
#   --------+----------------------+----------------+------------
#      X    |                      |                |     X (1)
#   --------+----------------------+----------------+------------
#           |          X           |                |
#   --------+----------------------+----------------+------------
#      X    |          X           |                |     X (1)
#   --------+----------------------+----------------+------------
#           |                      |        X       |
#   --------+----------------------+----------------+------------
#      X    |                      |        X       |
#   --------+----------------------+----------------+------------
#           |          X           |        X       |     X (2)
#   --------+----------------------+----------------+------------
#      X    |          X           |        X       |     X (2)
#   --------+----------------------+----------------+------------
#     (1) WAR to get around bug
#     (2) Replace programs handler with our own

sub _capture_warn_check {
	my $self = shift;

	if (!defined $SIG{__WARN__} ) {
		return $^V lt v5.8 ? 1 : 0;
	}
	return $self->{'FORCE_CAPTURE_WARN'} ? 1 : 0;
}
1;

__END__

#line 380


Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'any::feature',
                       'DISTNAME' => 'any-feature',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Backwards-compatible handling of new syntactic features',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100840',
                       'PREREQ_PM' => {
                                        'Scalar::Util' => '0',
                                        'Capture::Tiny' => '0',
                                        'English' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Perl::Version' => '0',
                                        'UNIVERSAL::require' => '0',
                                        'mysubs' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



